beacon-chain/sync/service.go

```
@@ -377,11 +377,12 @@ func (s *Service) waitForChainStart() {
 	}
 	s.ctxMap = ctxMap
 
-	// Register respective rpc handlers at state initialized event.
-	err = s.registerRPCHandlers()
-	if err != nil {
-		log.WithError(err).Error("Could not register rpc handlers")
-		return
+	// We need to register RPC handlers ASAP so that we can handle incoming status message
+	// requests from peers.
+	nse := params.GetNetworkScheduleEntry(clock.CurrentEpoch())
+	if err := s.registerRPCHandlers(nse); err != nil {
+		// If we fail here, we won't be able to peer with anyone because we can't handle their status messages.
+		panic(errors.Wrap(err, "Failed to register RPC handlers")) // lint:nopanic
```

is log.Fatal or panic preferred, i guess niether are preferred ...

---

is this change required vs the original code

---

If the node can't register RPC handlers here, it can't answer RPC status messages, which means it can't peer. So it really is a fatal error for the node. I can change it to log.Fatal, I guess the reason to do that would be consistent formatting in logs?

---

But I think the behavior of silently stopping this method while the node starts up is wrong, because this is a fatal condition for the node that requires operator intervention.

---

I think panic is not the right approach here. This could interrupt any in progress db transactions and cause other issues potentially. 

If this behavior needs to stop the client, can we surface the error up to the node and do a graceful shutdown ?

---

If that is too much refactoring for this PR, pls change back to log.Error and resolve in another PR. We shouldn't be panicking imo

---

Will do